File: kernel/src/driver/timers/rtc/rtc.c
#include "rtc.h"
#include <common/kprint.h>

/*置位0x70的第7位，禁止不可屏蔽中断*/

#define read_cmos(addr) ({                                                          \
    io_out8(0x70, 0x80 | addr);  \
    io_in8(0x71);                                                                   \
})

enum CMOSTimeSelector
{
    T_SECOND = 0x0,
    T_MINUTE = 0x2,
    T_HOUR = 0x4,
    T_DAY = 0x7,
    T_MONTH = 0x8,
    T_YEAR = 0x9,
};


int rtc_get_cmos_time(struct rtc_time_t *t)
{
    // 为防止中断请求打断该过程，需要先关中断
    cli();

    uint8_t status_register_B = read_cmos(0x0B);                  // 读取状态寄存器B
    bool is_24h = ((status_register_B & 0x02) ? true : false);    // 判断是否启用24小时模式
    bool is_binary = ((status_register_B & 0x04) ? true : false); // 判断是否为二进制码

    do
    {
        t->year = read_cmos(0x09);
        t->month = read_cmos(0x08);
        t->day = read_cmos(0x07);
        t->hour = read_cmos(0x04);
        t->minute = read_cmos(0x02);
        t->second = read_cmos(0x00);
    } while (t->second != read_cmos(0x00)); // 若读取时间过程中时间发生跳变则重新读取
    // 使能NMI中断
    io_out8(0x70, 0x00);

    if (!is_binary) // 把BCD转为二进制
    {
        t->second = (t->second & 0xf) + (t->second >> 4) * 10;
        t->minute = (t->minute & 0xf) + (t->minute >> 4) * 10;
        t->hour = ((t->hour & 0xf) + ((t->hour & 0x70) >> 4) * 10) | (t->hour & 0x80);

        t->month = (t->month & 0xf) + (t->month >> 4) * 10;
        t->year = (t->year & 0xf) + (t->year >> 4) * 10;
    }
    t->year += 2000;

    if ((!is_24h) && t->hour & 0x80) // 将十二小时制转为24小时
        t->hour = ((t->hour & 0x7f) + 12) % 24;
    sti();
    return 0;
}

File: kernel/src/driver/timers/rtc/rtc.rs
use crate::{
    arch::interrupt::{cli, sti},
    include::bindings::bindings::{io_in8, io_out8}, syscall::SystemError,
};

pub struct RtcTime {
    pub second: i32,
    pub minute: i32,
    pub hour: i32,
    pub day: i32,
    pub month: i32,
    pub year: i32,
}

impl Default for RtcTime {
    fn default() -> Self {
        Self {
            second: (0),
            minute: (0),
            hour: (0),
            day: (0),
            month: (0),
            year: (0),
        }
    }
}

impl RtcTime {
    ///@brief 从主板cmos中获取时间
    ///
    ///@param self time结构体
    ///@return int 成功则为0
    pub fn get(&mut self) -> Result<i32, SystemError> {
        // 为防止中断请求打断该过程，需要先关中断
        cli();
        //0x0B
        let status_register_b: u8 = read_cmos(0x0B); // 读取状态寄存器B
        let is_24h: bool = if (status_register_b & 0x02) != 0 {
            true
        } else {
            false
        }; // 判断是否启用24小时模式

        let is_binary: bool = if (status_register_b & 0x04) != 0 {
            true
        } else {
            false
        }; // 判断是否为二进制码

        loop {
            self.year = read_cmos(CMOSTimeSelector::Year as u8) as i32;
            self.month = read_cmos(CMOSTimeSelector::Month as u8) as i32;
            self.day = read_cmos(CMOSTimeSelector::Day as u8) as i32;
            self.hour = read_cmos(CMOSTimeSelector::Hour as u8) as i32;
            self.minute = read_cmos(CMOSTimeSelector::Minute as u8) as i32;
            self.second = read_cmos(CMOSTimeSelector::Second as u8) as i32;

            if self.second == read_cmos(CMOSTimeSelector::Second as u8) as i32 {
                break;
            } // 若读取时间过程中时间发生跳变则重新读取
        }

        unsafe {
            io_out8(0x70, 0x00);
        }

        if !is_binary
        // 把BCD转为二进制
        {
            self.second = (self.second & 0xf) + (self.second >> 4) * 10;
            self.minute = (self.minute & 0xf) + (self.minute >> 4) * 10;
            self.hour = ((self.hour & 0xf) + ((self.hour & 0x70) >> 4) * 10) | (self.hour & 0x80);
            self.day = (self.day & 0xf) + ((self.day / 16) * 10);
            self.month = (self.month & 0xf) + (self.month >> 4) * 10;
            self.year = (self.year & 0xf) + (self.year >> 4) * 10;
        }
        self.year += 2000;

        if (!is_24h) && (self.hour & 0x80) != 0 {
            self.hour = ((self.hour & 0x7f) + 12) % 24;
        } // 将十二小时制转为24小时

        sti();

        return Ok(0);
    }
}

///置位0x70的第7位，禁止不可屏蔽中断
#[inline]
fn read_cmos(addr: u8) -> u8 {
    unsafe {
        io_out8(0x70, 0x80 | addr);
        return io_in8(0x71);
    }
}

/// used in the form of u8
#[repr(u8)]
enum CMOSTimeSelector {
    Second = 0x00,
    Minute = 0x02,
    Hour = 0x04,
    Day = 0x07,
    Month = 0x08,
    Year = 0x09,
}
